teps/0026-interceptor-plugins.md

```diff
@@ -0,0 +1,510 @@
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+```go
+type InterceptorRequest struct {  
+	// Body is the incoming HTTP event body
+	Body []byte `json:"body,omitempty"`
+	// Header are the headers for the incoming HTTP event
+	Header map[string][]string `json:"header,omitempty"`
+
+	// InterceptorParams are the user specified params for the interceptor
+	InterceptorParams map[string]string `json:"interceptor_params,omitempty"`
+
+	Context *TriggerContext
+}
+
+type TriggerContext struct {
+	// EventUrl is the URL of the incoming event
+	EventUrl string `json:"url,omitempty"`
+	// EventID is a unique ID assigned by Triggers to each event
+	EventID string `json:"event_id,omitempty"`
+	// TriggerID is of the form namespace/$ns/triggers/$name
+	TriggerID string `json:"trigger_id,omitempty"`
+}
+```
+
+#### Response
+
+Interceptors in turn return a message with a fixed type:
+
+```go
+type InterceptorResponse struct {
+	// Extensions are  additional fields that is added to the interceptor event.
+	// See TEP-0022. Naming TBD.
+	Extensions map[string]interface{} `json:"extensions,omitempty"`
+	// Continue indicates if the EventListener should continue processing the Trigger or not
+	Continue bool `json:"continue,omitempty"`
+	// Status is an Error status t
+	Status grpc.Status `json:"status,omitempty"`
+}
+```
+
+#### Versioning
+
+To allow changes to the Interceptor HTTP interface, we can version the request and response types by embedding them in a `InterceptorMessage` type which contains an inline `TypeMeta`:
+
+```go
+type InterceptorMessage struct {
+    metav1.TypeMeta `json:",inline"`
+    // +optional
+    Request *InterceptorRequest `json:"request,omitempty" protobuf:"bytes,1,opt,name=request"`
+    // +optional
+    Response *InterceptorResponse `json:"response,omitempty" protobuf:"bytes,2,opt,name=response"`
+}
+```
+
+With this, an example JSON payload sent to the CEL Interceptor will look like the following:
+```json
+{
+  "apiVersion": "triggers.tekton.dev/v1alpha1",
+  "kind": "InterceptorMessage",
+  "request": {
+    "body": {}, // some body
+    "header": {
+      "eventType": "push",
+    }, 
+    "interceptor_params": {
+      "filter": "header.Match(\"eventType\",\"push\"")
+    },
+    "context": {
+      "event_id": "random-string",
+      "trigger_id": "namespaces/default/triggers/cel-trigger"
+    }
+  }
+}
+```
+And a sucessful response back will be as follows:
+
+```json
+{
+  "apiVersion": "triggers.tekton.dev/v1alpha1",
+  "kind": "InterceptorMessage",
+  "response": {
+    "continue": true,
+  }
+}
+```
+While a failed response will look like:
+
+```json
+{
+  "apiVersion": "triggers.tekton.dev/v1alpha1",
+  "kind": "InterceptorMessage",
+  "response": {
+    "continue": false,
+    "status": {
+      "code": "400", //Failed Precondition?
```

Comment:
@wlynch how do we use the GRPC status codes in a JSON body? Should the code be status.FAILED_PRECONDITION, or the HTTP equivalent code from the proto?

Reply:
https://developers.google.com/maps-booking/reference/grpc-api/status_codes